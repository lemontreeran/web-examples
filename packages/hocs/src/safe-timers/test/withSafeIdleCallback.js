/* eslint-disable max-len */
import React from 'react'
import { mount } from 'enzyme'

const Target = () => null

describe('withSafeIdleCallback', () => {
  let origRequestIdleCallback = null
  let origCancelIdleCallback = null
  let withSafeIdleCallback = null

  beforeAll(() => {
    origRequestIdleCallback = global.requestIdleCallback
    origCancelIdleCallback = global.cancelIdleCallback

    jest.resetModules()

    global.requestIdleCallback = jest.fn(() => 'id')
    global.cancelIdleCallback = jest.fn()

    withSafeIdleCallback = require('../src').withSafeIdleCallback
  })

  afterEach(() => {
    global.requestIdleCallback.mockClear()
    global.cancelIdleCallback.mockClear()
  })

  afterAll(() => {
    global.requestIdleCallback = origRequestIdleCallback
    global.cancelIdleCallback = origCancelIdleCallback
  })

  it('should pass props through', () => {
    const EnhancedTarget = withSafeIdleCallback(Target)
    const wrapper = mount(
      <EnhancedTarget a={1} b={2} />
    )

    expect(wrapper.find(Target)).toMatchSnapshot()
  })

  it('should provide `requestSafeIdleCallback` prop and unsubscriber as its call return', () => {
    const callback = () => {}
    const EnhancedTarget = withSafeIdleCallback(Target)
    const wrapper = mount(
      <EnhancedTarget />
    )
    const cancelSafeIdleCallback = wrapper.find(Target).prop('requestSafeIdleCallback')(callback, 'a', 'b')

    expect(global.requestIdleCallback.mock.calls).toMatchSnapshot()
    cancelSafeIdleCallback()
    expect(global.cancelIdleCallback.mock.calls).toMatchSnapshot()
  })

  it('should clear all safe intervals on unmount', () => {
    const EnhancedTarget = withSafeIdleCallback(Target)
    const wrapper = mount(
      <EnhancedTarget />
    )
    const requestSafeIdleCallback = wrapper.find(Target).prop('requestSafeIdleCallback')

    requestSafeIdleCallback()
    requestSafeIdleCallback()
    requestSafeIdleCallback()

    wrapper.unmount()

    expect(global.cancelIdleCallback.mock.calls).toMatchSnapshot()
  })

  describe('display name', () => {
    const origNodeEnv = process.env.NODE_ENV

    afterAll(() => {
      process.env.NODE_ENV = origNodeEnv
    })

    it('should wrap display name in non-production env', () => {
      process.env.NODE_ENV = 'test'

      const EnhancedTarget = withSafeIdleCallback(Target)
      const wrapper = mount(
        <EnhancedTarget />
      )

      expect(wrapper).toMatchSnapshot()
    })

    it('should not wrap display name in production env', () => {
      process.env.NODE_ENV = 'production'

      const EnhancedTarget = withSafeIdleCallback(Target)
      const wrapper = mount(
        <EnhancedTarget />
      )

      expect(wrapper).toMatchSnapshot()
    })
  })
})
