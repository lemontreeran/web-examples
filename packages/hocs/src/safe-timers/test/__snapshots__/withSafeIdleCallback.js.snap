// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withSafeIdleCallback display name should not wrap display name in production env 1`] = `
<SafeTimer>
  <Target
    requestSafeIdleCallback={[Function]}
  />
</SafeTimer>
`;

exports[`withSafeIdleCallback display name should wrap display name in non-production env 1`] = `
<withSafeIdleCallback(Target)>
  <Target
    requestSafeIdleCallback={[Function]}
  />
</withSafeIdleCallback(Target)>
`;

exports[`withSafeIdleCallback should clear all safe intervals on unmount 1`] = `
Array [
  Array [
    "id",
  ],
  Array [
    "id",
  ],
  Array [
    "id",
  ],
]
`;

exports[`withSafeIdleCallback should pass props through 1`] = `
<Target
  a={1}
  b={2}
  requestSafeIdleCallback={[Function]}
/>
`;

exports[`withSafeIdleCallback should provide \`requestSafeIdleCallback\` prop and unsubscriber as its call return 1`] = `
Array [
  Array [
    [Function],
    "a",
    "b",
  ],
]
`;

exports[`withSafeIdleCallback should provide \`requestSafeIdleCallback\` prop and unsubscriber as its call return 2`] = `
Array [
  Array [
    "id",
  ],
]
`;
